core/trino-main/src/main/java/io/trino/execution/buffer/OutputBufferMemoryManager.java

```diff
@@ -54,18 +56,29 @@
     @GuardedBy("this")
     private ListenableFuture<Void> blockedOnMemory = NOT_BLOCKED;
 
+    private final Ticker ticker = Ticker.systemTicker();
+
     private final AtomicBoolean blockOnFull = new AtomicBoolean(true);
 
     private final Supplier<LocalMemoryContext> memoryContextSupplier;
     private final Executor notificationExecutor;
 
+    @GuardedBy("this")
+    private final TDigest bufferUtilization = new TDigest();
+    @GuardedBy("this")
+    private long lastBufferUtilizationRecordTime;
+    @GuardedBy("this")
+    private double lastBufferUtilization;
+
     public OutputBufferMemoryManager(long maxBufferedBytes, Supplier<LocalMemoryContext> memoryContextSupplier, Executor notificationExecutor)
     {
         requireNonNull(memoryContextSupplier, "memoryContextSupplier is null");
         checkArgument(maxBufferedBytes > 0, "maxBufferedBytes must be > 0");
         this.maxBufferedBytes = maxBufferedBytes;
         this.memoryContextSupplier = Suppliers.memoize(memoryContextSupplier::get);
         this.notificationExecutor = requireNonNull(notificationExecutor, "notificationExecutor is null");
+        this.lastBufferUtilizationRecordTime = ticker.read();
+        this.lastBufferUtilization = getUtilization();
```

Comment:
It's bad practice to call instance methods from constructor (object is not fully initialized yet). You can remove this line as initial utilization will be 0